python/triton/compiler/compiler.py

```diff
@@ -21,20 +21,30 @@
 class AttrsDescriptor:
     divisible_by_16: set = None
     equal_to_1: set = None
+    within_2gb: set = None
+    non_negative: set = None
 
     def __post_init__(self):
         if self.divisible_by_16 is None:
             self.divisible_by_16 = set()
         if self.equal_to_1 is None:
             self.equal_to_1 = set()
+        if self.within_2gb is None:
+            self.within_2gb = set()
+        if self.non_negative is None:
+            self.non_negative = set()
 
     def to_dict(self):
-        return {'divisible_by_16': list(self.divisible_by_16), 'equal_to_1': list(self.equal_to_1)}
+        return {
+            'divisible_by_16': list(self.divisible_by_16), 'equal_to_1': list(self.equal_to_1), 'within_2gb':
+            list(self.within_2gb), 'non_negative': list(self.non_negative)
+        }
```

Comment:
let me have discuss that with people in OAI team for a bit (sorry didn't get to it so far). This is going increase the number of compilation quite a lot.

Reply:
May I ask what do you mean by _number of compilation_ ?

Reply:
every time we get a different specialization on a kernel we need to recompile it. Adding more specialization information will cause more kernels to be compiled

Reply:
So what you are saying is that `generic_kernel(-1, 2, -3)` and `generic_kernel(2,2,-3)` will have to be recompiled? I see. Right, I mean with @antiagainst we discussed that we could drop the non-negativeness properties for `tl.assume`s. This is not extremely generic, but might work for now. Later on we might introduce maybe an `increment_ptr` ptr operation that assumes (by definition) the offset being positive?

Reply:
I think the problem is smaller for `within_2gb`, right? But yes, I can also try to have per-backend specializations so that we don't touch other backends

Reply:
I think we can start with just `within_2gb` and pocket the `non_negative` one for now--we need to `tl.assume` a few scalars anyway due to otherwise hard to figure out; would rather just relying on `tl.assume` to provide the non negative info for now and see how that goes.